README.md

```diff
@@ -2,62 +2,62 @@
+1. Clone/download the [p4app](https://github.com/p4lang/p4app) GitHub repository `git clone https://github.com/p4lang/p4app`
+2. Append that repository to the `PATH` environment variable to let your console find the p4app executable. You can do this by something like this: `export PATH="/path/to/the/p4app/:$PATH"`
+3. Clone this repository containing our fastreact implementation `git clone https://github.com/steakhal/fastreact.p4app.git`
+4. Execute `p4app run .` in this folder. The application will shortly set up the environment, including the switch and the network.
+5. Open two separate WSL2 shells, for h1 and h2 nodes.
 6. Make sure those shells also have the `p4app` in the `PATH` envvar.
-7. Run the `p4app exec m h1 bash` command from one and the `p4app exec m h2 bash` in the other opened shell.
+7. Run `p4app exec m h1 bash` in the first and `p4app exec m h2 bash` in the second shell.
 8. Navigate to the `/tmp` folder in both shells.
-9. Run the `python3 send.py` in one, and the `python3 receive.py` in the other shell.
-
-You will probably see some packet communication.
-The last byte of the message represents the result of the rule matching logic on the switch.
+9. Run `python3 send.py` in the first and `python3 receive.py` in the second shell, whereas the first shell ships a packet towards the second. The last byte of the packet signs if the packet satisfied the switch-defined rule.
 
 ## Implementation logic
 
-The user can define rules in the following form:
+The user can define rules in conjunctive normal form:
 ```
 ((id00 op00 constant00)^(id01 op01 constant01)^(id02 op02 constant02)^(id03 op03 constant03))v
 ((id10 op10 constant10)^(id11 op11 constant11)^(id12 op12 constant12)^(id13 op13 constant13))v
 ((id20 op20 constant20)^(id21 op21 constant21)^(id22 op22 constant22)^(id23 op23 constant23))
 ```
-Each `id` is a sensor identifier, `op` a logical operation (except zero, which represents invalid operation) and a `constant` which we compare against.
-These (3x4x3) integer values represent a rule. As an example `4 2 7 0 0 0 0 0 0 0 0 0   0 0 0 0 0 0 0 0 0 0 0 0  0 0 0 0 0 0 0 0 0 0 0 0` represents the expression `id4 <= 7`.
-This rule is used in testing, where the rule table is populated by the `fastreact.config` file.
+where each triplet is consist of:
+- a sensor identifier (`id`)
+- a logical operation (`op`)
+- and a constant value (`constant`)
+
+Any triplet represent a literal, where we compare the sensor value to a constant (eg. S1 > 5, where S1 is the measured value by sensor 1)"
+Any rows (consist of 4 triplets) represent a condition, any of the triplets evaluating true satisfies the condtion (eg. S1 > 5 ∨ S2 = 8)
+The whole (3x4x3) integer values represent the rule, all of the conditions (represented by rows) needs to be fulfilled to satisfy the rule.
+(triplet starting with 0 is invalid, represents empty literal and evaluates true always)
```

Comment:
Whether a triplet is evaluated or not is driven only by the opcode.

The latter is more like a knowledgable comment relating to set theory. Which says that the empty set is part of any set.
In other words, the empty requirement set is fulfilled by any inputs.
=> Rule without any valid triplet is always evaluated to true regardless of the inputs.

Reply:
Fixed